code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,227 @@
+
+	var/list/cached_voices = list()
+
+	/// Whether TTS is enabled or not
+	var/tts_enabled = FALSE
+
+	var/message_timeout = 7 SECONDS
+
+	/// Messages can be timed out earlier if the algorithm thinks that
+	/// it's going to take too long for their message to be processed.
+	/// This'll determine the minimum extent of how late it is allowed to begin timing messages out
+	var/message_timeout_early_minimum = 5 SECONDS
+
+	var/max_concurrent_requests = 4
+
+	/// Used to calculate the average time it takes for a tts message to be received from the http server
+	/// For tts messages which time out, it won't keep tracking the tts message and will just assume that the message took
+	/// 7 seconds (or whatever the value of message_timeout is) to receive back a response.
+	var/average_tts_messages_time = 0
+
+/proc/cmp_word_length_asc(list/a, list/b)
+	return length(a[MESSAGE_INDEX]) - length(b[MESSAGE_INDEX])
+
+/datum/controller/subsystem/tts/vv_edit_var(var_name, var_value)
+	// tts being enabled depends on whether it actually exists
+	if(NAMEOF(src, tts_enabled) == var_name)
+		return FALSE
+	return ..()
+
+/datum/controller/subsystem/tts/stat_entry(msg)
+	msg = "Active:[length(in_process_tts_messages)]|Standby:[length(queued_tts_messages)]|Avg:[average_tts_messages_time]"
+	return ..()
+
+/datum/controller/subsystem/tts/Initialize()
+	if(!CONFIG_GET(string/tts_http_url))
+		return SS_INIT_NO_NEED
+
+	var/datum/http_request/request = new()
+	var/list/headers = list()
+	headers["Authorization"] = CONFIG_GET(string/tts_http_token)
+	request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/tts_http_url)]/tts-voices", "", headers)
+	request.begin_async()
+	UNTIL(request.is_complete())
+	var/datum/http_response/response = request.into_response()
+	if(response.errored || response.status_code != 200)
+		return SS_INIT_FAILURE
+	available_speakers = json_decode(response.body)
+	tts_enabled = TRUE
+	rustg_file_write(json_encode(available_speakers), "data/cached_tts_voices.json")
+	rustg_file_write("rustg HTTP requests can't write to folders that don't exist, so we need to make it exist.", "tmp/tts/init.txt")
+	return SS_INIT_SUCCESS
+
+/datum/controller/subsystem/tts/proc/play_tts(target, sound, datum/language/language, local, range = 7)
+	if(local)
+		SEND_SOUND(target, sound)
+		return
+
+	var/turf/turf_source = get_turf(target)
+	if(!turf_source)
+		return
+
+	var/channel = SSsounds.random_available_channel()
+	var/listeners = get_hearers_in_view(range, turf_source)
+
+	for(var/mob/listening_mob in listeners | SSmobs.dead_players_by_zlevel[turf_source.z])//observers always hear through walls
+		var/datum/language_holder/holder = listening_mob.get_language_holder()
+		if(!listening_mob.client?.prefs.read_preference(/datum/preference/toggle/sound_tts))
+			continue
+
+		if(get_dist(listening_mob, turf_source) <= range && holder.has_language(language, spoken = FALSE))
+			listening_mob.playsound_local(
+				turf_source,
+				sound,
```

Comment:
this isn't a valid input, the second arg is for text keys that match existing sounds. you shouldn't pass in a sound file as that. S redundant too